[promises] Filter unit test framework

ctiller opened the pull request:
Built atop #31448 

Offers a simple framework for testing filters.

<!--

If you know who should review your pull request, please assign it to that
person, otherwise the pull request would get assigned randomly.

If your pull request is for a specific language, please add the appropriate
lang label.

-->

markdroth commented on test/core/filters/filter_test.h:
```diff
@@ -0,0 +1,201 @@
+    void FinishNextFilter(ServerMetadataHandle md);
+
+    // Mock to trap starting the next filter in the chain.
+    MOCK_METHOD(void, Started, (const ClientMetadata& client_initial_metadata));
```
This seems like a bit of a weird use of gmock, since we're not actually mocking out any interface here.  In particular, this seems to push some boilerplate into the tests with the need to use things like `StrictMock<>`, and it requires the test author to understand the specific points at which gmock expectations are checked (generally not until the mock's destruction, which can sometimes be a bit counter-intuitive when reading the code).  It seems like maybe we'd be better off with a bespoke event notification system here, which the tests can check at the specific points where they expect the events to occur, maybe something like the [event queue](https://github.com/grpc/grpc/blob/49b5dfc14f2d3b8723dcc865e93c5a59f2538e4a/test/core/client_channel/lb_policy/lb_policy_test_lib.h#L431) that we use in the LB policy test framework?

ctiller replied:
I'd actually argue that the LB testing framework could use picking up something like this.

We seem to be in the process of designing and building an army of almost-gmock-matchers, each of which will be maintained and behave just a little bit differently.

Instead, I'd prefer to leverage already built and battle tested systems where I can, and save our teams efforts for building RPC systems instead of testing frameworks wherever possible.

Leveraging gmock here means I can use all of the actions built into gmock trivially, and have an already defined playbook for making extensions to the system.

markdroth commented on test/core/filters/client_auth_filter_test.cc:
```diff
@@ -140,72 +115,36 @@ TEST_F(ClientAuthFilterTest, CreateFailsWithoutRequiredChannelArgs) {
-      (*server_metadata)->get_pointer(GrpcMessageMetadata());
-  ASSERT_TRUE(message_md != nullptr);
-  EXPECT_EQ(message_md->as_string_view(), "access denied");
+  //  initial_metadata_batch_.Set(HttpAuthorityMetadata(), target_.Ref());
```
Please remove.

ctiller replied:
done

markdroth commented on test/core/filters/client_auth_filter_test.cc:
```diff
@@ -140,72 +115,36 @@ TEST_F(ClientAuthFilterTest, CreateFailsWithoutRequiredChannelArgs) {
+  call.Start(call.NewClientMetadata({{":authority", target()}}));
+  EXPECT_CALL(
+      call, Finished(AllOf(
+                HasMetadataKeyValue(
```
Failing a call with a non-OK status is likely to be a sufficiently common case that I would suggest adding a matcher just for that case.  It would be nice to say something like `CallStatusIs(absl::UnauthenticatedError("access denied"))`.

markdroth reviewed (commented):
This is really nice!  I like how it looks in the few tests we have, and it looks like it will make it very easy to write new tests.

My only substantive comment is the one about the use of gmock.  I'd really prefer to find an alternative for that.

Thanks!

ctiller commented:
Take a look and see what you think...

I added the verify to step like we talked about, but the verify is *per mocked object* and so that was going to need some call tracking in the previous scheme. Instead, I've moved all output from the test to a single mock type `Events`, and added an `EXPECT_EVENT` macro to queue stuff onto it.

I think it ends up in a really compelling place (and adding the base class we talked about would add unwarranted complexity here)

ctiller commented on test/core/filters/filter_test.h:
```diff
@@ -0,0 +1,201 @@
+  // Provides mocks to trap events that happen on the call.
+  class Call {
+   public:
+    ~Call();
```
done

ananda1066 commented on test/core/filters/filter_test.cc:
```diff
@@ -0,0 +1,421 @@
+
+void FilterTestBase::Call::Impl::Start(ClientMetadataHandle md) {
+  EXPECT_EQ(promise_, absl::nullopt);
+  promise_ = channel_->filter->MakeCallPromise(
```
This could use a comment.
